File: sdk/ts/src/tui/parity.ts
/**
 * @module tui/parity
 * @package @hop-top/kit
 *
 * Cross-language parity constants SoT. Values are read from the canonical
 * `contracts/parity/parity.json` at build time — tsup/esbuild inlines the
 * JSON into the emitted bundle, so the published package stays
 * self-contained without vendoring a second copy of the file.
 *
 * All TUI modules should import constants from here, not hardcode them.
 */

import parityRaw from '../../../../contracts/parity/parity.json';

/** Display metadata for a single help section. */
export interface SectionConfig {
  title: string;
}

/**
 * ParityData models every content block `parity.json` declares.
 *
 * A block present in the JSON but absent here is decoration: nothing loads
 * it, so no test and no consumer can see it drift. `PARITY_BLOCKS` below is
 * the machine-checkable half of that rule.
 */
export interface ParityData {
  description: string;
  status: {
    symbols: Record<'info' | 'success' | 'error' | 'warn', string>;
  };
  spinner: {
    frames: string[];
    interval_ms: number;
  };
  anim: {
    runes: string;
    interval_ms: number;
    default_width: number;
  };
  help: {
    /** Fang-vocabulary section names in render order. */
    section_order: string[];
    /** Display metadata keyed by fang section name. */
    sections: Record<string, SectionConfig>;
  };
  verbosity: {
    flag: string;
    /**
     * Maps the stacked `-V` count (as a decimal string) to a log level name.
     * Keys are strings because JSON object keys are.
     */
    levels: Record<string, string>;
    quiet_override: string;
  };
  streams: {
    flag: string;
    label_format: string;
    output: string;
  };
  /** Sibling contract files the parity suite also covers. */
  extends: string[];
}

/**
 * PARITY_BLOCKS is the registry of top-level `parity.json` keys this loader
 * knows. Every content block in `parity.json` MUST appear here and MUST have
 * a corresponding property on `ParityData` — the drift guard in
 * `parity.test.ts` enforces both directions so a new block cannot be added
 * as decoration.
 *
 * Mirrors `parity.Blocks` in `contracts/parity/parity.go`.
 *
 * Keys starting with `$` are JSON Schema metadata, not content.
 */
export const PARITY_BLOCKS = [
  'description',
  'status',
  'spinner',
  'anim',
  'help',
  'verbosity',
  'streams',
  'extends',
] as const;

export const parity: ParityData = parityRaw;

File: contracts/parity/parity.go
// Package parity provides cross-language TUI constants loaded from parity.json.
// This is the single source of truth for all TUI parity values (symbols,
// spinner frames, animation runes, timing, etc.) across Go, TypeScript, and Python.
//
// Other packages (tui, tui/styles) import from here; parity imports nothing
// from the kit tree to avoid circular dependencies.
package parity

import (
	_ "embed"
	"encoding/json"
)

//go:embed parity.json
var raw []byte

// SectionConfig holds display metadata for a single help section.
type SectionConfig struct {
	Title string `json:"title"`
}

// Data is the parsed content of parity.json.
type Data struct {
	Status struct {
		Symbols map[string]string `json:"symbols"`
	} `json:"status"`
	Spinner struct {
		Frames     []string `json:"frames"`
		IntervalMs int      `json:"interval_ms"`
	} `json:"spinner"`
	Anim struct {
		Runes        string `json:"runes"`
		IntervalMs   int    `json:"interval_ms"`
		DefaultWidth int    `json:"default_width"`
	} `json:"anim"`
	Help struct {
		SectionOrder []string                 `json:"section_order"`
		Sections     map[string]SectionConfig `json:"sections"`
	} `json:"help"`
	Verbosity struct {
		Flag string `json:"flag"`
		// Levels maps the stacked -V count (as a decimal string) to a
		// log level name. Keys are strings because JSON object keys are.
		Levels        map[string]string `json:"levels"`
		QuietOverride string            `json:"quiet_override"`
	} `json:"verbosity"`
	Streams struct {
		Flag        string `json:"flag"`
		LabelFormat string `json:"label_format"`
		Output      string `json:"output"`
	} `json:"streams"`
}

// Values is the parsed parity data, available at init time.
var Values Data

// Blocks is the registry of top-level parity.json keys this loader knows.
// Every content block in parity.json MUST appear here and MUST have a
// corresponding field on Data — TestParityNoUnloadedBlocks enforces both
// directions so a new block cannot be added as decoration.
//
// Keys starting with "$" are JSON Schema metadata, not content.
var Blocks = []string{
	"description",
	"status",
	"spinner",
	"anim",
	"help",
	"verbosity",
	"streams",
	"extends",
}

// Raw returns the embedded parity.json bytes. Tests use it to compare the
// declared key set against Blocks without re-reading the file from disk.
func Raw() []byte { return raw }

func init() {
	if err := json.Unmarshal(raw, &Values); err != nil {
		panic("tui/parity: failed to parse parity.json: " + err.Error())
	}
}
